controllers/managedfusiondeployment_controller.go

```diff
@@ -0,0 +1,446 @@
+			return fmt.Errorf("managed-fusion-agent-config secret contains an empty smtpPassword entry")
+		}
+
+		r.alertmanagerConfigSecret.Data = map[string][]byte{
+			pagerDutyKey:    []byte(r.pagerDutyConfigData.SecretKey),
+			smtpPasswordKey: []byte(r.smtpConfigData.Password),
+		}
+
+		return nil
+	})
+
+	return err
+}
+
+func (r *ManagedFusionDeploymentReconciler) reconcileAlertmanagerConfig() error {
+	r.Log.Info("Reconciling AlertmanagerConfig")
+
+	_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, r.alertmanagerConfig, func() error {
+		if err := r.own(r.alertmanagerConfig); err != nil {
+			return err
+		}
+
+		if r.pagerDutyConfigData.SOPEndpoint == "" {
+			return fmt.Errorf("managed-fusion-agent-config secret does not contain a SOPEndpoint entry")
+		}
+
+		alertingAddressList := []string{}
+		alertingAddressList = append(alertingAddressList,
+			r.smtpConfigData.NotificationEmails...)
+
+		if r.smtpConfigData.Username == "" {
+			return fmt.Errorf("managed-fusion-agent-config secret does not contain a username entry")
+		}
+		if r.smtpConfigData.Endpoint == "" {
+			return fmt.Errorf("managed-fusion-agent-config secret does not contain a endpoint entry")
+		}
+		if r.smtpConfigData.FromAddress == "" {
+			return fmt.Errorf("managed-fusion-agent-config secret does not contain a fromAddress entry")
+		}
+		smtpHTML, err := ioutil.ReadFile(r.CustomerNotificationHTMLPath)
+		if err != nil {
+			return fmt.Errorf("unable to read customernotification.html file: %v", err)
+		}
+
+		desired := templates.AlertmanagerConfigTemplate.DeepCopy()
+		for i := range desired.Spec.Receivers {
+			receiver := &desired.Spec.Receivers[i]
+			switch receiver.Name {
+			case "pagerduty":
+				receiver.PagerDutyConfigs[0].ServiceKey.Key = pagerDutyKey
+				receiver.PagerDutyConfigs[0].ServiceKey.LocalObjectReference.Name = r.alertmanagerConfigSecret.Name
+				receiver.PagerDutyConfigs[0].Details[0].Key = "SOP"
+				receiver.PagerDutyConfigs[0].Details[0].Value = r.pagerDutyConfigData.SOPEndpoint
+			case "SendGrid":
+				if len(alertingAddressList) > 0 {
+					receiver.EmailConfigs[0].Smarthost = r.smtpConfigData.Endpoint
+					receiver.EmailConfigs[0].AuthUsername = r.smtpConfigData.Username
+					receiver.EmailConfigs[0].AuthPassword.LocalObjectReference.Name = r.alertmanagerConfigSecret.Name
+					receiver.EmailConfigs[0].AuthPassword.Key = smtpPasswordKey
+					receiver.EmailConfigs[0].From = r.smtpConfigData.FromAddress
+					receiver.EmailConfigs[0].To = strings.Join(alertingAddressList, ", ")
+					receiver.EmailConfigs[0].HTML = string(smtpHTML)
+				} else {
+					r.Log.V(-1).Info("Customer Email for alert notification is not provided")
+					receiver.EmailConfigs = []promv1a1.EmailConfig{}
+				}
+			}
+		}
+		r.alertmanagerConfig.Spec = desired.Spec
+		utils.AddLabel(r.alertmanagerConfig, monLabelKey, monLabelValue)
+		return nil
+	})
+
+	return err
+}
+
+func (r *ManagedFusionDeploymentReconciler) removeOLMComponents() error {
```

Comment:
A better name at this point should be
```suggestion
func (r *ManagedFusionDeploymentReconciler) initiateAgentUninstallation() error {
```